compiler/symbol.go

```diff
@@ -354,6 +361,44 @@ func (info *functionInfo) parsePragmas(f *ssa.Function) {
 	}
 }
 
+// Check whether this function cannot be used in //go:wasmimport. It will add an
+// error if this is the case.
+//
+// The list of allowed types is based on this proposal:
+// https://github.com/golang/go/issues/59149
+func (c *compilerContext) checkWasmImport(f *ssa.Function, pragma string) {
+	if c.pkg.Path() == "runtime" {
+		// The runtime is a special case. Allow all kinds of parameters
+		// (importantly, including pointers).
+		return
+	}
+	if f.Blocks != nil {
+		// Defined functions cannot be exported.
+		c.addError(f.Pos(), fmt.Sprintf("can only use //go:wasmimport on declarations"))
+		return
+	}
+	for _, param := range f.Params {
+		typ := param.Type().Underlying()
+		// Check whether the type is allowed.
+		// Only a very limited number of types can be mapped to WebAssembly.
+		allowedType := false
+		switch typ := typ.(type) {
+		case *types.Basic:
+			switch typ.Kind() {
+			case types.Int32, types.Uint32, types.Int64, types.Uint64:
+				allowedType = true
+			case types.Float32, types.Float64:
+				allowedType = true
+			case types.UnsafePointer:
+				allowedType = true
+			}
+		}
+		if !allowedType {
+			c.addError(param.Pos(), fmt.Sprintf("%s: unsupported parameter type %s", pragma, param.Type().String()))
+		}
+	}
+}
```

Comment:
In Go, we also validate the return values (see https://github.com/golang/go/blob/86c6b4763ed486f20cf018d0810cfd7a1fd91998/src/cmd/compile/internal/ssagen/abi.go#L392-L415)

Note the key difference is we don't allow returning pointers, only integer/floats.

Reply:
Thanks, I missed return values! I've updated the PR accordingly.

Also thanks for mentioning pointers, because I almost missed those.

Reply:
@achille-roussel is who I would ask, so I approve indirectly ;)

Reply:
but heh, I should mention that some signatures are using unsafe pointers. There was that chat about nil vs zero recently. So, are we saying that folks should coerce to uintptr(0) in the rare case the signature accepted an unsafe pointer? https://github.com/tinygo-org/tinygo/blob/0d56dee00f49bd50eb373c02c30062a75ec28f10/src/runtime/arch_tinygowasm_malloc.go#L43

Reply:
uintptr isn't supported actually because unlike the rest of the industry, GOARCH=wasm uses 64 bits pointers so we didn't want to risk weird bugs by implicitly truncating values when crossing import boundaries. 

I guess the answer to "how do I return a pointer to a region that the host allocated in the linear memory?" is to return a uint32. It's not great but we decided it would be better to push that complexity to the code rather than pretending it can be handled safely by the compiler. 

Baby steps :)

Reply:
fwiw this works for me. At first, I found pointer types semantically helpful, but from a mapping perspective disturbing. I didn't realize GOARCH=wasm uses 64 bits pointers either! I kindof assumed it went with wasm 1 and 2 which don't support large memory.

Reply:
> but heh, I should mention that some signatures are using unsafe pointers.

The example you give is from the runtime, which is special-cased. The restrictions on `//go:wasmimport` don't apply to the runtime.

> uintptr isn't supported actually because unlike the rest of the industry, GOARCH=wasm uses 64 bits pointers so we didn't want to risk weird bugs by implicitly truncating values when crossing import boundaries.

...and TinyGo uses 32-bit `uintptr` (because I didn't check when first adding support for wasm, I just assumed it would be 32-bits). Not allowing `uintptr` in `//go:wasmimport` does in fact avoid potential issues there.

Reply:
Thanks for the clarification!